libs/timeserieslib/TimeSeriesValidator.h

```diff
@@ -0,0 +1,176 @@
+#ifndef TIMESERIESVALIDATOR_H
+#define TIMESERIESVALIDATOR_H
+
+#include "TimeSeries.h"
+
+
+using boost::posix_time::time_duration;
+
+namespace mkc_timeseries
+{
+    class TimeSeriesValidationException : public std::runtime_error
+    {
+    public:
+        TimeSeriesValidationException(const std::string msg)
+            : std::runtime_error(msg)
+        {}
+
+        ~TimeSeriesValidationException()
+        {}
+    }; 
+
+    template <class Decimal>
+    class TimeSeriesValidator
+    {
+        public:
+            TimeSeriesValidator(
+                std::shared_ptr<OHLCTimeSeries<Decimal>> hourlyTimeSeries, 
+                std::shared_ptr<OHLCTimeSeries<Decimal>> dailyTimeSeries) :
+            mHourlyTimeSeries(hourlyTimeSeries), 
+            mDailyTimeSeries(dailyTimeSeries) 
+            {}
+
+            void validate() 
+            {
+                mValidateSevenTimePeriods();
+                mValidateTimeStamps(); 
+                mValidateAvailableDays();
+            }
+
+        private:
+            std::shared_ptr<OHLCTimeSeries<Decimal>> mHourlyTimeSeries;
+            std::shared_ptr<OHLCTimeSeries<Decimal>> mDailyTimeSeries;
+
+            void mValidateTimeStamps() // warning
+            {
+                boost::posix_time::time_duration typicalStartTime = boost::posix_time::hours(9);
+                boost::posix_time::time_duration typicalEndTime = boost::posix_time::hours(15);
+                for(auto it = mHourlyTimeSeries->beginRandomAccess(); it != mHourlyTimeSeries->endRandomAccess(); it++)
+                {
+                    if(it->getBarTime() < typicalStartTime || it->getBarTime() > typicalEndTime)
+                        std::cout << "WARNING: Bar time " << it->getBarTime() << " is out of the typical trading time range (9:00 - 15:00)" << std::endl;
+                }
+            }
+
+            void mValidateSevenTimePeriods() // error 
+            {
+                boost::gregorian::date startDate = mHourlyTimeSeries->beginRandomAccess()->getDateValue();
+                std::vector<boost::posix_time::time_duration> distinctTimeDurations;
+                int sevenDayCount = 0, holidayCount = 0, numberDays = 0;
+                std::vector<boost::gregorian::date> daysToDelete;
+                bool validateTimeDurations = true;
+                for(auto it = mHourlyTimeSeries->beginRandomAccess(); it != mHourlyTimeSeries->endRandomAccess(); it++)
+                {
+                    if(it->getDateValue() != startDate) 
+                    {
+                        if(!isEarlyCloseDay(startDate))
+                        {
+                            if(distinctTimeDurations.size() == 7)    // distinct time frames per day
```

Comment:
Anthony would it be possible to generalize the check for "7" into a check for the number time frames we "should have"? This code will work fine for equities, but if we ever try this on futures it will fail.

Reply:
Yeah, not sure why I didn't think to do that. I will grab the number of time frames from the timeframe inference procedure that is done right before this validation.